Completes initial version

ludanin opened the pull request:
I added the compiled code to main so we can only review the differences in src at this PR

Commit:
Completes initial version

Commit:
Fixes font creation

Commit:
Fixes links weight

Commit:
Exports ThemeProvider

Commit:
Fixes body's navbar padding/margin

tianhuil commented on src/style/color.ts:
```diff
@@ -0,0 +1,51 @@
+   * Used throughout the app, often a tame color when compared to `accent`
+   * in order to avoid overwhelming users with visual pollution.
+   */
+  readonly primary: Complete extends true ? ColorSwatch : Partial<ColorSwatch>
```
Why not do this?
```
ColorThemeBase<ColorSwatch=string>
```
So that the colors are just a string by default but you could pass in whatever color swatch you want?  This makes it more flexible so that we're future proof.

tianhuil commented on src/style/theme.ts:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  /**
+   * Generic global style to be applied on the page, defaults to fixes,
```
I like this

tianhuil commented on src/components/TextInput.tsx:
```diff
@@ -0,0 +1,37 @@
+      | 'url'
+  }
+
+export const TextInput: React.FC<Props> = (props) => {
```
Let's rename this `LabelInput`.  Does it have to be for text or can it be for numbers as well?

ludanin replied:
It can be used for numbers, but also we can omit the label if we want. Maybe this should be only `Input`, and then create Components for all the other `input.type`, e.g. `Radio` (for HTML `<input type='radio'/>`), `Checkbox`, etc.

tianhuil commented on src/style/defaultTheme.ts:
```diff
@@ -0,0 +1,303 @@
+} from '.'
+
+const normalizeMin =
+  '/*! normalize.css v8.0.1 | MIT License | github.com/necolas/normalize.css */html{line-height:1.15;-webkit-text-size-adjust:100%}body{margin:0}main{display:block}h1{font-size:2em;margin:.67em 0}hr{box-sizing:content-box;height:0;overflow:visible}pre{font-family:monospace,monospace;font-size:1em}a{background-color:transparent}abbr[title]{border-bottom:none;text-decoration:underline;text-decorati ...
```
+1!

tianhuil commented on src/components/Navbar.tsx:
```diff
@@ -0,0 +1,231 @@
+  Declaring these as CSS variables since they change according to devices'
+  breakpoints.
+  */
+  :root {
```
I don't really understand but could we not make these variables part of the Theme to be specified?  For example under `Navbar` part of the theme?

ludanin replied:
The main reason I put them here is that they might change as device orientation changes (or if users resize their window). It's possible to achieve the same result using Theme, but that would mean we'd have to do some manual tasks:

  * Keep track of the window dimension using events (or `useScreenClass`), and change `height` accordingly;
  * We'd also have to allow ourselves to rewrite the logic for calculating `expandedHeight` with a callback in other projects using the package, it's way easier to do this using CSS (only a `calc(height * amountOfLinks)`);
  * Currently when the `Navbar` loads we put a margin to `body` so no content touches the sidebar, if the Sidebar is not present this margin defaults to `0` due to the nature of CSS Variables, it's possible to do this with JS--but again we'd have to check if the user is on mobile/desktop for the different `navHeights`);

I think that if we continue with CSS variables for the Navbar then I should doc-comment that these variables are exported, and with an example of why they are exported globally in CSS.

tianhuil replied:
My experience has been that `useScreenClass` is a little buggy (it's always `xs` on server-side rendering (i.e. the first loaded page) so let's avoid that.

tianhuil commented on src/components/Navbar.tsx:
```diff
@@ -0,0 +1,231 @@
+  css?: SerializedStyles | SerializedStyles[]
+}
+
+export const Navbar: React.FC<Props> = ({ links, children, css }) => {
```
Since we already have to have multiple ones, why don't we call the `children` `logo` and not have any children?

tianhuil commented on src/style/font.ts:
```diff
@@ -0,0 +1,29 @@
+import { ThemeCSSRule } from '.'
+
+export interface FontTheme {
+  /** Defaults to sans-serif */
```
I don't understand `sans` / `serif` / `mono`.  We never use it.  Shouldn't `text` / `h1` - `h6` be enough?

ludanin replied:
I'm using these mostly to avoid errors, TypeScript or ESLint will not complain if we have
```css
  font-family: "Open San", sans-serif; /* Note the typo on Sans */
```
So whenever we need to refer to font family I'm using the values from the theme. Another feature of these is that by default headers are going to use `font.accent`, and the page is going to use `font.sans` for text. These values are easy to overwrite using `font.text`/`font.h1h6`.

tianhuil replied:
I agree about having defined constants to mean font names to avoid spellings.  I just don't think that all websites will follow the scheme `sans` / `serif` / `mono` / `accent` etc ...

Instead, let's just call it `Record<string, string>` and let the websites subclass what `font` needs to be.

tianhuil commented on src/style/grid.ts:
```diff
@@ -0,0 +1,11 @@
+import { styled } from '.'
+export { ContainerProps, RowProps, ColProps } from 'react-grid-system'
+
+export const Container = styled(_Container)()
```
+1!

tianhuil commented on src/style/makeTheme.ts:
```diff
@@ -0,0 +1,135 @@
+}
+
+export const makeTheme = (theme?: Partial<ThemeBase<false>>): Theme => {
+  // This ternary operation allow us to avoid `TypeError: theme is undefined`
```
Sorry -- why don't we just create these by hand?  Why do we need the `makeTheme` helper?  Happy to just chat about it.

ludanin replied:
makeTheme currently reduces the amount of work required to create a theme, for example most of our applications use a default light theme, instead of having us manually copying-pasting all the color values, then the radius, then all the rules for h1-h6, makeTheme will provide the default values in case we don't want to customize these.

Another feature `makeTheme` has is the ability to allow overwriting the default rules, for example we already have rules for h1-h6, but an application (say `virtual-coffee` for example) has all these in `color.primary.normal`. We can keep the default rules for h1-h6, and call makeTheme with:
```tsx
  const theme = makeTheme({
    font: {
      h1h6: {
        replace: false, // Keeps the original style, we don't want to clear it
        css: ({color}) => css`
          color: ${color.primary.normal};
        `,
      },
    },
  })
```

Note that in this case we didn't even declare the rules for colors, radius, buttons, inputs, etc. We wanted only to change the colors for headers. `makeTheme` in this example would call another helper function which would in turn merge the `defaultTheme` for `h1h6` with the new provided one.

tianhuil commented on src/style/radius.ts:
```diff
@@ -0,0 +1,16 @@
+ * a smaller one for checkboxes, small buttons, tooltips, and a bigger one for components
+ * that are often detached from the page, e.g. modals, cards.
+ */
+export interface RadiusTheme {
```
+1

tianhuil commented on src/style/util.ts:
```diff
@@ -0,0 +1,86 @@
+  typeof num === 'number' ? `${num}px` : num
+
+/**
+ * Parses colors, the purpose of this function is to make all 3 digits HEX
```
Is this needed?  Emotion seems to understand `#fff` already.

Commit:
Fixes headers global theme

Commit:
Several small fixes

Commit:
Another set of fixes

Commit:
Exports props

Commit:
Improves rollup & build, exports awesome reveal

tianhuil commented on src/inspectlet.js:
```diff
@@ -0,0 +1,7 @@
+function inspectlet(inspectletId) {
```
Why not make this a `*.ts` and declare the typing here.

tianhuil commented on src/style/color.ts:
```diff
@@ -0,0 +1,38 @@
+ * In other words, PartialColorTheme allows not only for `Partial<Color>`
+ * but also for each `Color[type]` to be `Partial<ColorSwatch>`
+ */
+export type PartialColorTheme = Partial<ColorThemeBase<false>>
```
This is not necessary since `ColorThemeBase<false>` already has a `Partial` surrounding it.

tianhuil commented on src/style/font.ts:
```diff
@@ -0,0 +1,29 @@
+  /** Defaults to serif */
+  readonly serif: string
+  /** Defaults to monospace */
+  readonly mono: string
```
I would just make FontTheme a `Record<string, string>` and let the individual webpage decide.  For example, a webpage may decide to use `{text: 'xxx', header: 'yyy'}` or just `{text: 'xxx'}` or something more complex.  I would not go about customizing it this way.

tianhuil commented on src/style/makeTheme.ts:
```diff
@@ -0,0 +1,105 @@
+      accent: font?.accent
+        ? `"${font.accent}", sans-serif`
+        : defaultTheme.font.accent,
+      mono: font?.mono ? `"${font.mono}", monospace` : defaultTheme.font.mono,
```
We don't need this complex section for font.

tianhuil commented on src/style/Provider.tsx:
```diff
@@ -0,0 +1,83 @@
+          }
+        `}
+      />
+      <Global
```
We don't need all the global sections for fonts.  Instead, I would just take a `cssCss` argument for where we can implement whatever we want for font (or anything else).  Basically, I don't think giving font any special structure is helpful.

tianhuil commented on src/style/Provider.tsx:
```diff
@@ -0,0 +1,83 @@
+  return (
+    <Fragment>
+      <Global styles={theme.global.css(theme)} />
+      <Global styles={theme.font.text.css(theme)} />
```
We don't need all the global sections for fonts (i.e. text / h1 - h6).  Instead, I think we can stick with `theme.global.css` as sufficient.  Or we can use the above global.css Basically, I don't think giving font any special structure is helpful.

tianhuil commented on src/style/defaultTheme.ts:
```diff
@@ -0,0 +1,302 @@
+      from these rules
+      */
+      html {
+        color: ${color.ink.normal};
```
I'd remove anything that has too much to do with font and leave it to whatever normalize chooses  It's more clear / better documented to define them in the individual websites.

tianhuil reviewed (changes requested):
Some more changes.  The main thing is to simplify how we are handling font (make more powerful and less useful) and push that code (usefulness) to the websites (see #1).  Where we do err on the side of being more useful (less powerful), I would choose a well-established convention (see #2).

1. Let's make `Theme` declare a type variable `FontTheme extends Partial<Record<string, string>>` which takes on values like this:
```ts
interface FontTheme: {
    text: string
    header: string
} 
```
or
```ts
interface FontTheme: {
    normal: string
    accent1: string
    accent2: string
} 
```
or whatever the specific site wants.  But let's keep it flexible for the site.

2. I'm OK with `ColorSwatch` -- it looks like you're mimicking https://material-ui.com/customization/color/ and https://material-ui.com/customization/palette/

If so, let's change the names to be consistent: (light / main / dark) and call it `palette` not color.  We would want to use
```
interface PaletteColor {
  light: string;
  main: string;
  dark: string;
  contrastText: string;
}
```
so that we remain consiste.

And I would stick to only using these predefined colors (i.e. no accent):
```
palette.primary
palette.secondary
palette.error
palette.warning
palette.info
palette.success
```

Commit:
Uses typed Inspectlet

Commit:
Removes FontTheme, uses normalize for h1h6

Exports Google GA

Commit:
Exports ReactGA

Commit:
Comments nabvar

Commit:
Adds Modal & Checkbox; Lot of fixes & improvements

Adds Form & Loading to be used with it (automatically hides content
when `loading === true`)

Fixes Button not passing all props on the global style.

Renames TextInput => Input

Exports Visible & Hidden from grid

Adds more props to CommonStyleProps

ludanin commented:
Added some code improvements & support for modal, checkbox, and a `<Form/>` Component that automatically hides content while displaying a spinning loading indicator when `loading === true`.

For more info check: 57bb689

Commit:
Implements Radio

Commit:
Implements TextArea

Commit:
Reworks props for ThemeCSSFunc & Modal Props

Moves animateMS from Theme to ModalProps

Commit:
Exports untyped styled

Commit:
Reworks typings to allow extendable theme

Commit:
Fixes Modal
